stackslib/src/clarity_vm/clarity.rs

```diff
@@ -294,6 +295,26 @@ impl ClarityInstance {
             )
         };
 
+        // If we're starting a new tenure, and we're in epoch 3.09, then we
+        // need to increment the tenure height in the Clarity DB.
+        if new_tenure && epoch.epoch_id >= StacksEpochId::Epoch30 {
```

Comment:
This seems brittle to me.  If you're going to introduce a new MARF key/value pair anyway for representing the tenure height, can you instead update it as part of the Nakamoto chain state's `append_block` function?

Reply:
I need the value while building the block, for use while evaluating the transactions included in the block. If we do it in `append_block`, then I think it will be off by 1.

Reply:
That won't happen?  You can read an inserted a key/value pair into the MARF before committing it, so if you insert the new tenure height before processing Clarity transactions, you should be fine.

Reply:
Ah, okay. I was thinking `append_block` was called after the transactions were processed. I'm a little unclear where exactly I would need to put this and where to get the needed information. I can dig into that tomorrow, but if you've got something in mind, the help would be appreciated.

Reply:
I think the best strategy is to add this to `setup_block` (which gets invoked in both the miner path and the follower path). Something like this should work:

```diff
diff --git a/stackslib/src/chainstate/nakamoto/mod.rs b/stackslib/src/chainstate/nakamoto/mod.rs
index 2dbcd9f3c..5a9a4b2be 100644
--- a/stackslib/src/chainstate/nakamoto/mod.rs
+++ b/stackslib/src/chainstate/nakamoto/mod.rs
@@ -2551,6 +2551,28 @@ impl NakamotoChainState {
             new_tenure,
         );
 
+        if new_tenure {
+            clarity_tx
+                .connection()
+                .as_free_transaction(|clarity_tx_conn| {
+                    clarity_tx_conn.with_clarity_db(|db| {
+                        let tenure_ht = db.get_tenure_height()?;
+                        db.set_tenure_height(
+                            tenure_ht
+                                .checked_add(1)
+                                .expect("FATAL: tenure height overflowed"),
+                        )?;
+                        Ok(())
+                    })
+                })
+                .map_err(|e| {
+                    error!("Failed to set tenure height during block setup";
+                           "error" => ?e,
+                    );
+                    e
+                })?;
+        }
+
         // now that we have access to the ClarityVM, we can account for reward deductions from
         // PoisonMicroblocks if we have new rewards scheduled
         let matured_rewards_opt = matured_rewards_schedule_opt
```

Then you wouldn't need to pass `new_tenure` through the block begin methods.

Reply:
Thanks for this @kantai. Do you think the initialization in `initialize_epoch_3_0` good to stay there?

Reply:
Moved in 008f73bf4e5926aa1c8b2c747a06014e37e9e102.

Reply:
Better yet, you can just call `db.set_tenure_height(coinbase_height)` since it ought to equal to the height of the current tenure.  No need to re-compute a value that's already given!  (cc @kantai for confirmation)

Reply:
Yeah -- `coinbase_height` should be the same value.